replications/remotewrite/writer.go

```diff
@@ -85,91 +86,83 @@ func NewWriter(replicationID platform.ID, store HttpConfigStore, metrics *metric
+	}
 
-	for {
+	res, postWriteErr := PostWrite(ctx, conf, data, w.clientTimeout)
+	res, msg, ok := normalizeResponse(res, postWriteErr)
+	if !ok {
+		// bail out
+		return w.backoff(attempts), true, postWriteErr
+	}
 
-		// Get the most recent config on every attempt, in case the user has updated the config to correct errors.
-		conf, err := w.configStore.GetFullHTTPConfig(ctx, w.replicationID)
-		if err != nil {
-			return err
-		}
+	// Update metrics and most recent error diagnostic information.
+	if err := w.configStore.UpdateResponseInfo(ctx, w.replicationID, res.StatusCode, msg); err != nil {
+		// TODO: We shouldn't fail/retry a successful remote write for not successfully writing to the config store
+		// we should log instead of returning, like:
+		// w.logger.Warn("failed to update config store with latest remote write response info", zap.Error(err))
+		// Unfortunately this will mess up a lot of tests that are using UpdateResponseInfo failures as a proxy for
+		// write failures.
+		return w.backoff(attempts), true, err
+	}
 
-		res, err := PostWrite(ctx, conf, data, w.clientTimeout)
-		res, msg, ok := normalizeResponse(res, err)
-		if !ok {
-			// Can't retry - bail out.
-			return err
-		}
+	if postWriteErr == nil {
+		// Successful write
+		w.metrics.RemoteWriteSent(w.replicationID, len(data))
+		w.logger.Debug("remote write successful", zap.Int("attempt", attempts), zap.Int("bytes", len(data)))
+		return 0, true, nil
+	}
 
-		// Update metrics and most recent error diagnostic information.
-		if err := w.configStore.UpdateResponseInfo(ctx, w.replicationID, res.StatusCode, msg); err != nil {
-			return err
-		}
+	w.metrics.RemoteWriteError(w.replicationID, res.StatusCode)
+	w.logger.Debug("remote write error", zap.Int("attempt", attempts), zap.String("error message", "msg"), zap.Int("status code", res.StatusCode))
 
-		if err == nil {
-			// Successful write
-			w.metrics.RemoteWriteSent(w.replicationID, len(data))
-			w.logger.Debug("remote write successful", zap.Int("attempt", attempts), zap.Int("bytes", len(data)))
-			return nil
-		}
+	var waitTime time.Duration
+	hasSetWaitTime := false
 
-		w.metrics.RemoteWriteError(w.replicationID, res.StatusCode)
-		w.logger.Debug("remote write error", zap.Int("attempt", attempts), zap.String("error message", "msg"), zap.Int("status code", res.StatusCode))
-
-		attempts++
-		var waitTime time.Duration
-		hasSetWaitTime := false
-
-		switch res.StatusCode {
-		case http.StatusBadRequest:
-			if conf.DropNonRetryableData {
-				w.logger.Debug("dropped data", zap.Int("bytes", len(data)))
-				w.metrics.RemoteWriteDropped(w.replicationID, len(data))
-				return nil
-			}
-		case http.StatusTooManyRequests:
-			headerTime := w.waitTimeFromHeader(res)
-			if headerTime != 0 {
-				waitTime = headerTime
-				hasSetWaitTime = true
-			}
+	switch res.StatusCode {
+	case http.StatusBadRequest:
+		if conf.DropNonRetryableData {
+			w.logger.Debug("dropped data", zap.Int("bytes", len(data)))
```

Comment:
* this should be `warn` level debugging I think- dropped data is serious.
* maybe we can print the error message from the body in the log as well?